### PR TITLE
(856b) Edit/update participation course

### PR DESCRIPTION
**Pull request by yndajas**
## Context

We can currently add a new course participation by selecting a course, then access the participation details page, but we can't edit the course for a course participation

## Changes in this PR

- Allow editing/updating a course participation's course

This PR has a couple of significant abstractions and some minorly complicated `describe.each` work going on, but hopefully the order and content of the commits makes them understandable. Happy to do a run through of the PR if it would help the review process!

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [x] There are changes required to the Accredited Programmes API for this change to work...
  - [x] ... and they have been released to production already

### Post-merge

- [ ] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [x] This makes new expectations of the API...
  - [x] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

**Comment by yndajas** [comment]
[Add page for editing a previous programme's name (M)](https://trello.com/c/At6EXbEV/856-add-page-for-editing-a-previous-programmes-name-m)

File: server/views/referrals/courseParticipations/course.njk [hunk 1]
```diff
@@ -40,7 +41,7 @@
         }) }}
       {% endif %}
 
-      <form action="{{ referPaths.programmeHistory.create({ referralId: referralId }) }}" method="post">
+      <form action="{{ action }}" method="post">
```
Comment (jsrobertson) [inline server/views/referrals/courseParticipations/course.njk:44]: 👍

File: server/controllers/refer/courseParticipationsController.ts [hunk 2]
```diff
@@ -104,12 +143,58 @@ export default class CourseParticipationsController {
...
+  updateCourse(): TypedRequestHandler<Request, Response> {
+    return async (req: Request, res: Response) => {
+      TypeUtils.assertHasUser(req)
+
+      const currentCourseParticipation = await this.courseService.getParticipation(
+        req.user.token,
+        req.params.courseParticipationId,
+      )
+
+      if (!currentCourseParticipation) {
```
Comment (jsrobertson) [inline server/controllers/refer/courseParticipationsController.ts:167]: This should already be handled in the service.

File: server/utils/courseParticipationUtils.test.ts [hunk 3]
```diff
@@ -11,6 +15,68 @@ const getRow = (
...
+          expect(CourseParticipationUtils.processedCourseFormData(otherId, '  ', request)).toEqual({
+            hasFormErrors: true,
+            otherCourseName: '',
+          })
+          expect(request.flash).toHaveBeenCalledWith('otherCourseNameError', 'Enter the programme name')
+        })
+      })
+    })
+
+    describe('when the `courseId` is `undefined``', () => {
```
Comment (Gweaton) [inline server/utils/courseParticipationUtils.test.ts:71]: Double ``` at the end here (sorry!)

File: server/utils/courseParticipationUtils.test.ts [hunk 4]
```diff
@@ -11,6 +15,68 @@ const getRow = (
...
+          courseId: validId,
+          hasFormErrors: false,
+        })
+      })
+    })
+
+    describe("when the `courseId` is `'other'`", () => {
+      const otherId = 'other'
+
+      describe('and `otherCourseName` is a string with length when trimmed', () => {
```
Comment (Gweaton) [inline server/utils/courseParticipationUtils.test.ts:40]: maybe "not an empty string when trimmed" would be a bit more human-readable here.

**Review by Gweaton — APPROVED** [review]
This looks good - that abstraction seems to have simplified the tests a bit!

Just a couple of minor wording tweaks.

File: server/controllers/refer/courseParticipationsController.ts (hunk 2 above)
Comment (yndajas) [reply server/controllers/refer/courseParticipationsController.ts:167]: Ah, I thought you'd handled it somewhere!

File: server/utils/courseParticipationUtils.test.ts (hunk 4 above)
Comment (yndajas) [reply server/utils/courseParticipationUtils.test.ts:40]: I think I lost the will to word here

File: server/controllers/refer/courseParticipationsController.ts (hunk 2 above)
Comment (yndajas) [reply server/controllers/refer/courseParticipationsController.ts]: I've updated the `getParticipation` return type [here](https://github.com/ministryofjustice/hmpps-accredited-programmes-ui/pull/245/commits/1ed0e84196e9fe7294f1ff5ff9290fe1dcbe41ed) - hope that's correct!

File: server/utils/courseParticipationUtils.test.ts (hunk 3 above)
Comment (yndajas) [reply server/utils/courseParticipationUtils.test.ts]: Good spot!

**Pull request merged**